[feat] New performance syntax

jjotero opened the pull request:
This PR introduces a new syntax to capture and report performance data from a reframe check. In order to facilitate writing generic and composable tests, the `reference` variable is now entirely optional and each performance variable is extracted with a member function decorated with the `performance_function` decorator:
```python
import reframe as rfm
import reframe.utility.sanity as sn
import re

@rfm.simple_test
class Base(rfm.RunOnlyRegressionTest):
    valid_systems = ['dom:login']
    valid_prog_environs = ['PrgEnv-gnu']
    executable = 'echo 40 bananas'
    v = variable(str)

    @sanity_function
    def assert_bananas(self): 
        return sn.assert_found(r'bananas', self.stdout)
         
    @performance_function('bananas')
    def perf0(self):
        return sn.extractsingle(r'(\d+)\s*bananas', self.stdout, 1, int)

    @performance_function('bananas', perf_key='b-1')
    def perf1(self, blah=222):
        f = open(self.stdout.evaluate(), 'r')
        stdout = ''.join(f.readlines())
        f.close()
        return int(re.match(r'(\d+)\s*bananas', stdout)[1])
```
Reframe will pickup these functions marked as `@performance_functions` and run them all by default:
```bash
==============================================================================
PERFORMANCE REPORT
------------------------------------------------------------------------------
Base
- dom:login
   - PrgEnv-gnu
      * num_tasks: 1
      * perf0: 40 bananas
      * b-1: 40 bananas
------------------------------------------------------------------------------
```
Note that the units are attached to the decorated function, and the name of the resulting performance variable can also be customised with the `perf_key` decorator argument. If any decorated performance function has more than one argument without a default value (the `self` placeholder), an error is raised:
```python
def foo(x, y=10):
    # cool

def bar(x, y):
   # not cool
```
So now, what if someone derives a test and wants to change the performance variables from the base test? We have the `perf_variables` attribute for that. During class instantiation, ReFrame will initialise `perf_variables` with all the functions decorated with the `performance_function` decorator. However, if one wants to modify this mapping, one can do so in a similarly to when dealing with the old `perf_patterns` as:
```python
@rfm.simple_test
class Derived(Base):
    @run_before('performance')
    def my_report(self):
        l = lambda self: sn.extractsingle(r'(\d+)\s*bananas', self.stdout, 1, int)
        self.perf_variables = {
                'a': self.perf0(),
                'b': sn.make_performance_function(l, 'more_bananas', self),
                'c': sn.make_performance_function(
                    sn.extractsingle(r'(\d+)\s*bananas', self.stdout, 1, int),
                    'other_bananas'
                )
        }
```
With this syntax, the functions that go into the `perf_variables` dictionary are guaranteed to be suitable performance functions, otherwise an error is raised. Note the addition of the `sn.make_performance_function` method, which allows to add more performance functions inline (it works with both a generic callable and a deferred expression). The performance report for this `Derived` test looks as follows:
```bash
==============================================================================
PERFORMANCE REPORT
------------------------------------------------------------------------------
Base
- dom:login
   - PrgEnv-gnu
      * num_tasks: 1
      * a: 40 bananas
      * b: 40 more_bananas
      * c: 40 other_bananas
------------------------------------------------------------------------------
```

Closes #2064
Closes #2113

pep8speaks commented:
Hello @jjotero, Thank you for submitting the Pull Request!









* In the file [`reframe/core/pipeline.py`](https://github.com/eth-cscs/reframe/blob/7607dea19e9c09c32ee53c0471322064a4ab3d92/reframe/core/pipeline.py):

> [Line 589:26](https://github.com/eth-cscs/reframe/blob/7607dea19e9c09c32ee53c0471322064a4ab3d92/reframe/core/pipeline.py#L589): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent



Do see the ReFrame [Coding Style Guide](https://github.com/eth-cscs/reframe/wiki/coding-style-guide)

jgphpc commented on cscs-checks/microbenchmarks/gpu/gpu_burn/gpu_burn_test.py:
```diff
@@ -24,30 +24,29 @@ class gpu_burn_check(GpuBurn):
     reference = {
         'dom:gpu': {
-            'perf': (4115, -0.10, None, 'Gflop/s'),
+            'min_perf': (4115, -0.10, None),
```
```suggestion
            my_limits = "-0.10, None, 'Gflop/s'"
            ...
            'min_perf': (4115, my_limits),
```
Feel free to ignore

jgphpc replied:
nevermind

jjotero replied:
I see what you mean there. I guess that one can still have the units as a comment in there instead of being part of the reference tuple. At the moment setting the reference tuple the "old way" when using this new syntax would throw an error. I think it's a good thing to throw an error, because otherwise people would use a bit of a hybrid syntax, which might lead to errors if at some point we drop the current syntax

victorusu reviewed (commented):
@jjotero, I really loved this PR! This syntax is very clear, extensible, and very powerful. 
But I have a question why is the `make_performance_function` a regression test method? Would it make sense to have it as an independent module that can be used by regression classes? I think that the module approach is more "modular".
And last but not least, I loved, really loved the bananas examples! I do not know if the `b-1` example was intentional, but I missed a [`b-2`](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjMpPXBlvHxAhWx_7sIHX8KDmUQyCkwAHoECA0QAw&url=https%3A%2F%2Fwww.youtube.com%2Fwatch%3Fv%3DHHxQUw0Upfk&usg=AOvVaw2mip0UDm_FdtpjgG6vdxgd)!

jjotero commented:
@victorusu I was not aware of that show! 😂
`self.make_performance_function` is implemented as a wrapper of the `@performance_function` decorator, which is not available from the instance. Where else are you thinking this `make_performance_function` method would be used if not in a regression test instance?

jgphpc commented on reframe/core/pipeline.py:
```diff
@@ -1608,52 +1609,101 @@ def check_performance(self):
+            if hasattr(self, '_rfm_perf_report') or self._rfm_perf_fns:
+                if hasattr(self, 'perf_patterns'):
+                    raise ReframeSyntaxError(
+                        f"assigning a value to 'perf_patters' conflicts with ",
```
typo

jgphpc commented on reframe/core/meta.py:
```diff
@@ -252,6 +253,69 @@ def sanity_function(fn):
+
+        # Machinery to add performance variables
+        def performance_function(units, *, perf_key=None):
+            '''Decoreate a function to extract a performance variable.
```
typo

jgphpc commented on reframe/core/pipeline.py:
```diff
@@ -1608,52 +1609,101 @@ def check_performance(self):
+                    self._perf_logger.log_performance(logging.INFO, tag, value,
+                                                      *self.reference[key])
+
+            # Check the performace variables against their references.
```
typo

jgphpc reviewed (approved):
lgtm
[gist](https://gist.github.com/jgphpc/1483c954053a32c2e2f4517a447026f7) to myself.

ekouts reviewed (approved):
Lgtm so far. After all agreeing on the syntax we should update the documentation

vkarak commented:
I've started looking into the new syntax. I like the concept of the `@performance_function` and the way it works. One of my first comments is that I'm generally against removing the units from the reference. When I read a test and I see the units I can immediately get an understanding of the measurements. Otherwise I would have to look into the various performance functions, which can be scattered over different files. I'm generally against on the recommendation of writing the unit as a comment, because very easily it'll become outdated. I'd vote to issue a warning if the unit in the perf function and the reference do not match. That'd be helpful for example when a library changes; if the unit of a performance function changes, then most likely the reference must be updated.

Regarding the `@performance_report` I'm still looking at it :-)

jjotero commented:
Changed the `@performance_report` decorator to `@performance_variables`.

vkarak commented:
@brandongc If I understand correctly your point, that could be an option, but we didn't want to touch the `perf_patterns` syntax for backward compatibility. Also, as a matter of style we believe the decorators feel a bit more natural and integrated to the language compared to something like this:

```python
perf_patterns = [PerfVar(...), PerfVar(...)]
```

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -1655,52 +1656,111 @@ def check_performance(self):
+                # even if the don't meet the reference.
+                for tag, expr in self.perf_patterns.items():
+                    value = sn.evaluate(expr)
+                    key = '%s:%s' % (self._current_partition.fullname, tag)
```
We can convert this to an `f-string` as well as some others further down.

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -1655,52 +1656,111 @@ def check_performance(self):
+
+                            # Pop the unit from the ref tuple (redundant)
+                            ref = ref[:3]
+
```
Let's remove this line.

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -1655,52 +1656,111 @@ def check_performance(self):
+                for tag, expr in self.perf_patterns.items():
+                    try:
+                        value, unit = sn.evaluate(expr)
+                        unit = getattr(unit, '_rfm_perf_units', None)
+                        if not unit:
+                            raise TypeError('Trigger the TypeError below')
+
+                    except TypeError:
```
This pattern is a bit weird here. We get a  unit that actually may not contain a unit?

vkarak reviewed (changes requested):
I finished the review of the PR since my comments are not related to what we decide on the `@performance_variables` decorator.

jjotero commented on reframe/core/namespaces.py:
```diff
@@ -59,7 +59,7 @@ def __repr__(self):
         '''Raise an error if there is a namespace clash.'''
-        raise ValueError(
+        raise KeyError(
             f'{name!r} is already present in the current namespace'
```
Yes, I was thinking though, that perhaps it makes sense to leave that one as a `KeyError` since this class just implements a base namespace and the users are never exposed to this low level namespace mechanisms. Thoughts?

vkarak replied:
Agreed. This is not the final exception thrown to the user. Let's leave it as such.

jjotero reviewed (commented):
I have introduced the variable `perf_variables` of type `typ.Dict[str, _DeferredPerformanceExpression]`. A `_DeferredPerformanceExpression` is derived from the `_DeferredExpression` class with the addition that it has a member `unit` holding the performance units of each performance variable.

As a result of this internal reorg, I have moved the `make_performance_function` utility into the pipeline. This function is basically an external convert-constructor for the `_DeferredPerformanceExpression` class that can take either a callable with its arguments and the units, or just a deferred expression and the units. This could be integrated somehow into the typed descriptor, and it probably makes sense to think of this also in relation to #563 (which also needs some type of conversion function).

jjotero commented:
I've added the unit tests for all this new syntax. Also, I've moved the `make_performance_function` method from the pipeline into the `utility/sanity` module, since this method can be used with a free function. Now, since this `make_performance_function` returns a `_DeferredPerformanceExpression`, I think we should be consistent on how we name this function and all the other `deferrable` and so on. I'll try to think of a better name for this.

vkarak commented on reframe/core/deferrable.py:
```diff
@@ -56,17 +56,17 @@ def evaluate(self):
             )
 
         ret = self._fn(*fn_args, **fn_kwargs)
-        if isinstance(ret, type(self)):
+        while isinstance(ret, _DeferredExpression):
```
Perhaps a comment here that we use `while` to capture recursive deferrable calls would be helpful.

vkarak commented on cscs-checks/microbenchmarks/gpu/gpu_burn/gpu_burn_test.py:
```diff
@@ -62,17 +61,31 @@ def set_gpu_arch(self):
+                    val = self.perf(nid).evaluate()[0]
+                    sn.assert_reference(val, ref, lt, ut).evaluate()
+                except SanityError:
+                    perf_report[nid] = self.perf(nid)
```
This sounds like a good idea.

vkarak commented on reframe/utility/__init__.py:
```diff
@@ -304,6 +304,27 @@ def attrs(obj):
     return ret
 
 
+def is_trivially_callable(fn, *, non_def_args=0):
```
Yes, I figured this and it's useful, but I suggest to adapt a bit the documentation of the function because now it reads as "A trivially callable object is one that can be called without providing any additional arguments to its call method," which is a bit contradictory 😭

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -1655,52 +1656,111 @@ def check_performance(self):
+                    raise ReframeSyntaxError(
+                        f"assigning a value to 'perf_pattenrs' conflicts ",
+                        f"with using the 'performance_variables' decorator ",
+                        f"(class {self.__class__.__qualname__})"
```
> The only thing is that internally it would help if we have a single representation as we do now.

I don't remember what I exactly I meant with that, but I agree with the current design.

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -1666,63 +1691,52 @@ def check_performance(self):
-                        f"with using the 'performance_variables' decorator ",
+                        f"with using the 'performance_function' decorator ",
+                        f"or setting a value to 'perf_variables' "
                         f"(class {self.__class__.__qualname__})"
```
The name of the class is not needed here. It will be printed by the loader.

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -1666,63 +1691,52 @@ def check_performance(self):
+                    except Exception as e:
+                        logging.getlogger().warning(
+                            f'skipping evaluation of performance variable '
+                            f'{tag!r} in test {self.name!r}: {e}'
```
Again, we don't need the test name here. The logger knows the context of the test, so it could log the test name regardless.

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -1666,63 +1691,52 @@ def check_performance(self):
+                                    f'reference unit ({key!r}) for the '
+                                    f'performance variable {tag!r} '
+                                    f'does not match the unit specified '
+                                    f'in the performance function ({unit!r})'
```
Perhaps we need to say which one of the two will be ignored. We can simply add that "{unit!r} will be used"

vkarak commented on unittests/test_pipeline.py:
```diff
@@ -1249,6 +1249,134 @@ def extract_perf(patt, tag):
+    # Make the unit in the reference different from the performance function
+    perftest.reference = {
+        '*': {
+            'value1': (0, None, None, 'unit_')
```
I think that you could assert that the warning about the unit mismatch in `log_output`.

vkarak replied:
You're right. You can't! I thought we were using the logger 😅

vkarak commented on unittests/test_pipeline.py:
```diff
@@ -1249,6 +1249,134 @@ def extract_perf(patt, tag):
+
+def test_perf_function_raises_exception(perftest, sanity_file,
+                                        perf_file, dummy_gpu_exec_ctx):
+    # The lambda takes a single arg, and the make_performance_function
+    # will inject the self argument too. This triggers an exception
```
Is this accurate?

vkarak commented on reframe/core/meta.py:
```diff
@@ -271,7 +261,7 @@ def sanity_function(fn):
 
-        # Machinery to add performance variables
+        # Machinery to add performance functions
         def performance_function(units, *, perf_key=None):
```
We still need to document this separately in the rst file.

vkarak commented on unittests/test_meta.py:
```diff
@@ -254,3 +256,91 @@ class AllowFinalOverride(Base):
+    Base().assert_perf_fn_return()
+
+    # Test the performance function set has been built correctly
+    assert len(Base._rfm_perf_fns) == 2
+    perf_fns = {'perf_a', 'perf_b'}
+    for fn in Base._rfm_perf_fns:
+        assert fn._rfm_perf_key in perf_fns
+        perf_fns.remove(fn._rfm_perf_key)
+
+    assert not perf_fns
```
I think the following covers everything:

```suggestion
    assert {fn._rfm_perf_key for fn in Base._rfm_perf_fns} == {'perf_a', 'perf_b'}
```

vkarak commented on unittests/test_meta.py:
```diff
@@ -254,3 +256,91 @@ class AllowFinalOverride(Base):
+            '''Perf function with custom key.'''
+
+    # Test the performance function set is correct with class inheritance
+    assert len(Derived._rfm_perf_fns) == 3
+    perf_fns = {'perf_b', 'perf_c', 'my_perf_fn'}
+    for fn in Derived._rfm_perf_fns:
+        assert fn._rfm_perf_key in perf_fns
+        perf_fns.remove(fn._rfm_perf_key)
+
+    assert not perf_fns
```
Check my suggestion above.

vkarak commented on unittests/test_meta.py:
```diff
@@ -254,3 +256,91 @@ class AllowFinalOverride(Base):
+
+
+def test_performance_function_errors(MyMeta):
+    with pytest.raises(ReframeSyntaxError):
```
I have the feeling that here we can throw a `TypeError` and the error would still be traced back to the right user line. Similarly for anything in the `performance_function` decorator.

vkarak replied:
> Changed to TypeError, but now I'm confused 😅. I thought the plan was to raise ReframeSyntaxError for everything that we impose and control.

If a programming error can be traced back to the user frame that triggered it, then ReFrame can avoid a stack trace automatically. In this case, it can be traced because it's a direct call chain from the decorator call. The problem has been with the variables and parameters, where we do several checks at a later point. In these cases, the user line information is lost and we can't trace back the error to the user code; the stack trace has no user frame. So to avoid printing a stack trace for such user errors, we invented the `ReframeSyntaxError`. That's the rule of thumb, I hope it makes sense...

vkarak replied:
I don't know where would be the right place to be honest, because that's mostly a developer's documentation, isn't it?

vkarak commented on hpctestlib/microbenchmarks/gpu/gpu_burn/__init__.py:
```diff
@@ -96,27 +96,27 @@ def count_successful_burns(self):
+        if metric not in {'perf', 'temp'}:
             raise ValueError(
-                f"unsupported value in 'what' argument ({what}!r)"
+                f"unsupported value in 'metric' argument ({metric}!r)"
```
```suggestion
                f"unsupported value in 'metric' argument: {metric}!r"
```

vkarak commented on reframe/core/meta.py:
```diff
@@ -82,7 +82,13 @@ def __setitem__(self, key, value):
+                    self['_rfm_perf_fns'][key] = value
+                except KeyError:
+                    raise ReframeSyntaxError(
+                        f'the performance function {key} has already been '
```
```suggestion
                        f'the performance function {key!r} has already been '
```

vkarak replied:
I'm fixing that now.

vkarak commented on reframe/core/parameters.py:
```diff
@@ -45,6 +46,16 @@ def filter_params(x):
+        try:
+            valid = utils.is_trivially_callable(filter_params, non_def_args=1)
+        except TypeError:
+            raise TypeError('the provided parameter filter is not a callable')
```
```suggestion
            raise TypeError('the provided parameter filter is not a callable') from None
```

vkarak replied:
I'm fixing that now.

vkarak commented on docs/deferrable_functions_reference.rst:
```diff
@@ -48,6 +50,18 @@ Currently ReFrame provides three broad categories of deferrable functions:
    They include, but are not limited to, functions to iterate over regex matches in a file, extracting and converting values from regex matches, computing statistical information on series of data etc.
 
 
+Deferrable performance functions
```
I would make this a subsection

vkarak commented on docs/deferrable_functions_reference.rst:
```diff
@@ -48,6 +50,18 @@ Currently ReFrame provides three broad categories of deferrable functions:
+Deferrable function utilities
+-----------------------------
+
 Users can write their own deferrable functions as well.
 The page ":doc:`deferrables`" explains in detail how deferrable functions work and how users can write their own.
```
This paragraph should probably be moved before the "performance functions."

vkarak commented on docs/deferrable_functions_reference.rst:
```diff
@@ -48,6 +50,18 @@ Currently ReFrame provides three broad categories of deferrable functions:
+Regular deferrable functions can be promoted to deferrable performance functions using the :func:`~reframe.utility.sanity.make_performance_function` utility.
+Also, this utility allows to create performance functions directly from any callable.
+
+Deferrable function utilities
```
```suggestion
List of deferrable function and utilities
```

vkarak commented on docs/regression_test_api.rst:
```diff
@@ -322,6 +322,21 @@ Built-in functions
+   In the event of a name conflict, the derived class will follow Python's `MRO <https://docs.python.org/3/library/stdtypes.html#class.__mro__>`_ to choose the appropriate performance function.
+   However, defining more than one performance function with the same name in the same class is disallowed.
+
+   The full set of performance functions of a regression test is stored under :attr:`~reframe.core.pipeline.RegerssionTest.perf_variables` as key-value pairs, where, by default, the key is the name of the decorated member function, and the value is the deferred performance function itself.
```
```suggestion
   The full set of performance functions of a regression test is stored under :attr:`~reframe.core.pipeline.RegressionTest.perf_variables` as key-value pairs, where, by default, the key is the name of the decorated member function, and the value is the deferred performance function itself.
```

vkarak replied:
I'm fixing that now.

vkarak commented on docs/regression_test_api.rst:
```diff
@@ -322,6 +322,21 @@ Built-in functions
+
+   Decorate a member function as a performance function of the test.
+
+   This decorator converts the decorated method into a performance deferrable function (see :ref:`deferrable-functions`) whose evaluation is deferred to the performance stage of the regression test.
```
We can have a more precise link here to the actual subsection.

vkarak commented on reframe/core/deferrable.py:
```diff
@@ -66,10 +67,15 @@ def evaluate(self):
 
+        # Cache the results for any subsequent evaluate calls.
         self._cached = (ret,)
+
```
Remove this line.

vkarak reviewed (changes requested):
Looks good! Just minor corrections and some fine tuning to the docs. I could also take care of some of those.

vkarak commented on reframe/core/pipeline.py:
```diff
@@ -647,22 +647,36 @@ def pipeline_hooks(cls):
+    #:
+    #: In this context, a performance variable is a key-value pair, where the
+    #: key is the desired variable name and the value is the deferred
+    #: performance expression (i.e. the result of a :doc:`deferrable
```
Here I suggest adding `.. _deferrable-perf-functions:` before the section in the sanity functions documentation. Then you can reference it anywhere with ``:ref:`deferrable-perf-functions` `` and the title of the section will be inserted. To avoid typing here, I would also change this to ``see :ref:`..` for more information`.

vkarak reviewed (approved):
Latm! Goes in! I just made a bit of fine tuning of the docs.

vkarak commented on reframe/utility/sanity.py:
```diff
@@ -904,15 +910,22 @@ def defer(x):
+    If ``expr`` is a deferred expression and ``cache`` is ``True``, the
+    results of the deferred expression will be cached and subsequent calls
+    to ``evaluate`` on this deferred expression will simply return the
+    previously cached result.
```
Yeah, it was too late yesterday night that I was reading it. Now that I reread it's clear also from this sentence :-) The additional note doesn't hurt though.
